boa_engine/src/syntax/parser/expression/primary/mod.rs

```diff
@@ -126,17 +126,47 @@ where
                 ClassExpression::new(self.name, self.allow_yield, self.allow_await)
                     .parse(cursor, interner)
             }
-            TokenKind::Keyword((Keyword::Async, false)) => {
-                cursor.next(interner).expect("token disappeared");
-                let mul_peek = cursor.peek(1, interner)?.ok_or(ParseError::AbruptEnd)?;
-                if mul_peek.kind() == &TokenKind::Punctuator(Punctuator::Mul) {
-                    AsyncGeneratorExpression::new(self.name)
-                        .parse(cursor, interner)
-                        .map(Node::from)
-                } else {
-                    AsyncFunctionExpression::new(self.name, self.allow_yield)
+            TokenKind::Keyword((Keyword::Async, contain_escaped_char)) => {
+                let contain_escaped_char = *contain_escaped_char;
+                let is_generator = match cursor.peek(2, interner) {
+                    Ok(Some(token)) => token.kind() == &TokenKind::Punctuator(Punctuator::Mul),
+                    _ => false,
+                };
+
+                // FIXME: even if the next character is OpenParen, it could be an identifier.
+                // e.g.) async();
+                match cursor.peek(1, interner) {
+                    Ok(Some(token)) => match token.kind() {
+                        TokenKind::Keyword((Keyword::Function, _))
+                        | TokenKind::Punctuator(Punctuator::OpenParen)
+                            if contain_escaped_char =>
+                        {
+                            let tok = cursor.peek(0, interner)?.ok_or(ParseError::AbruptEnd)?;
+                            Err(ParseError::general(
+                                "Keyword must not contain escaped characters",
+                                tok.span().start(),
+                            ))
+                        }
+                        TokenKind::Keyword((Keyword::Function, _))
+                        | TokenKind::Punctuator(Punctuator::OpenParen) => {
```

Comment:
Why are we looking for OpenParen here? That would be the start of a CallExpression or an AsyncArrowFunction and neither of those are parsed in PrimaryExpression right?

I think when we have implemented the parsing of AsyncArrowFunction this pattern should probably never be valid here, because it would be parsed earlier.

Reply:
I looked into it again, and it seems like I was looking at wrong code. yea, you're right. this check should've done in other place. Thanks for catching this!